runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -796,71 +797,87 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+      if (arrayConstraint &&
+          isArrayElementFlattened(arrayConstraint) == TR_yes &&
+          fe()->getComponentClassFromArrayClass(arrayConstraint->getClass()))
          {
-         flags8_t flagsForTransform(isLoadFlattenableArrayElement ? ValueTypesHelperCallTransform::IsArrayLoad
-                                                                  : ValueTypesHelperCallTransform::IsArrayStore);
-         flagsForTransform.set(ValueTypesHelperCallTransform::InsertDebugCounter);
-
-         if (isStoreFlattenableArrayElement && !owningMethodDoesNotContainStoreChecks(this, node))
-            {
-            // If storing to an array whose component type is or might be a value type
-            // and the value that's being assigned is or might be null, both a run-time
-            // NULLCHK of the value is required (guarded by a check of whether the
-            // component type is a value type) and an ArrayStoreCHK are required;
-            // otherwise, only the ArrayStoreCHK is required.
-            //
-            if ((isCompTypeVT != TR_no) && (storeValueConstraint == NULL || !storeValueConstraint->isNonNullObject()))
-               {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreAndNullCheck);
-               }
-            else
-               {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
-               }
-            }
-
+         flags16_t flagsForTransform(isLoadFlattenableArrayElement ? ValueTypesHelperCallTransform::IsFlattenedArrayLoad
+                                                                   : ValueTypesHelperCallTransform::IsFlattenedArrayStore);
          if (!owningMethodDoesNotContainBoundChecks(this, node))
             {
             flagsForTransform.set(ValueTypesHelperCallTransform::RequiresBoundCheck);
             }
 
          _valueTypesHelperCallsToBeFolded.add(
-               new (trStackMemory()) ValueTypesHelperCallTransform(_curTree, node, flagsForTransform));
+               new (trStackMemory()) ValueTypesHelperCallTransform(_curTree, node, flagsForTransform, arrayConstraint->getClass()));
          }
       else
          {
-         // Construct static debug counter to note failure to take advantage of
-         // any VP constraint to eliminate this array element helper call
-         const char *operationName = isLoadFlattenableArrayElement ? "aaload" : "aastore";
+         // If the array's component type is definitely not a value type, or if the value
+         // being assigned in an array store operation is definitely not a value type, add
+         // a delayed transformation to replace the helper call with inline code to
+         // perform the array element access.
+         if ((arrayConstraint != NULL && isCompTypeVT == TR_no)
+             || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
+            {
```

Comment:
I think we had discussed adapting this to handle the cases where the array is known to be of a value type that is not flattened, so that a reference can simply be loaded or stored without a helper call.  Should that case be handled as part of this pull request or as a follow on improvement?

Reply:
Sorry, my bad, I'll make the change as a follow up improvement.

Reply:
Addressed in 24b108d